File: Infrastructure/automation-service/app/schemas/calendar.py
"""Pydantic schemas for calendar API."""

from __future__ import annotations

from datetime import date
import os
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, validator


class CalendarEventCreate(BaseModel):
    location: str
    cluster: str = "main"
    event_type: str = "planned_task"
    title: str
    start_date: date
    end_date: date | None = None
    all_day: bool = True
    notes: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class CalendarEventUpdate(BaseModel):
    title: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    notes: str | None = None
    event_type: str | None = None
    location: str | None = None
    cluster: str | None = None
    metadata: dict[str, Any] | None = None


class FlowerGrowPlanRequest(BaseModel):
    idempotency_key: UUID
    crop_name: str = "Flower cycle"
    environment: str = "indoor"
    flower_end: date
    flower_weeks: int = Field(ge=1, le=52)
    include_pot_phases: bool = True
    clone_weeks: int = Field(default=3, ge=1, le=12)
    pot_weeks: int = Field(default=2, ge=1, le=12)
    bed_weeks: int = Field(default=2, ge=1, le=12)
    stretch_days: int = Field(default=21, ge=1, le=60)
    ripen_days: int = Field(default=21, ge=1, le=60)
    drying_days: int = Field(default=7, ge=1, le=14)
    auto_mode_transition: bool = True


class SyncConnectionCreate(BaseModel):
    display_name: str | None = None
    account_email: str | None = None
    caldav_base_url: str
    username: str
    app_password: str
    target_calendar_url: str


class SyncConnectionTest(BaseModel):
    caldav_base_url: str
    username: str
    app_password: str

    @validator("caldav_base_url")
    @classmethod
    def require_https(cls, v: str) -> str:
        if v.startswith("http://"):
            if os.getenv("CALDAV_ALLOW_HTTP_TEST", "").lower() != "true":
                raise ValueError(
                    "CALDAV connection test requires HTTPS. Set CALDAV_ALLOW_HTTP_TEST=true for dev only."
                )
        return v

File: Infrastructure/frontend/src/types/calendar.ts
/** Calendar event DTO from automation API. */
export interface CalendarEventDto {
  id: string;
  source: 'manual' | 'mode_transition' | 'crop_batch';
  eventType: string;
  title: string;
  start: string;
  end?: string;
  location: string;
  cluster?: string;
  editable: boolean;
  colorKey: string;
  numericId?: number;
  notes?: string;
  metadata?: Record<string, unknown>;
}

export interface CalendarRoomProfile {
  location: string;
  display_name: string;
  color_key: string;
  sort_order: number;
}

export interface CalendarEventsResponse {
  items: CalendarEventDto[];
  next_cursor?: string | null;
}

export interface FlowerGrowPlanRequest {
  idempotency_key: string;
  crop_name: string;
  environment: 'indoor' | 'outdoor';
  flower_end: string;
  flower_weeks: number;
  include_pot_phases: boolean;
  clone_weeks?: number;
  pot_weeks?: number;
  bed_weeks?: number;
  stretch_days?: number;
  ripen_days?: number;
  drying_days?: number;
  auto_mode_transition?: boolean;
}

export interface ModeScheduleResponse {
  date: string;
  expected: {
    mode_name: string | null;
    submode_name: string | null;
    event_type?: string;
    title?: string;
  };
  active: {
    mode_name: string | null;
    submode_name: string | null;
  };
}

export const ROOM_CALENDAR_COLORS: Record<string, string> = {
  'Flower Room': 'bg-amber-500',
  'Veg Room': 'bg-emerald-500',
  Lab: 'bg-slate-500',
};

export const EVENT_TYPE_COLORS: Record<string, string> = {
  clone_window: 'bg-blue-400',
  pot_veg: 'bg-blue-500',
  bed_veg: 'bg-blue-600',
  flip_to_flower: 'bg-amber-400',
  flower_stretch: 'bg-pink-400',
  flower_bulk: 'bg-pink-500',
  flower_ripen: 'bg-pink-600',
  drying: 'bg-amber-600',
  harvest: 'bg-rose-500',
  mode_transition: 'bg-purple-500',
  planned_task: 'bg-cyan-500',
};
